### PR TITLE
kubernetes-helm: 3.1.2 -> 3.1.2 and 2.16.5 split to helm 2 and 3

### PR DESCRIPTION
**Pull request by dredozubov**
###### Motivation for this change

Helm 2 and 3 are architecturally different and there are lots of projects still using helm 2. It makes sense to be able to build the last version of Helm 2 as well as Helm 3.

###### Things done

* split `kubernetes helm` to `kubernetes-helm2` and `kubernetes-helm3`
* default `kubernetes-helm` to `kubernetes-helm3` for compatibility

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---
This change is [Reviewable](https://reviewable.io/reviews/nixos/nixpkgs/83219)

### CODE REVIEW

**Review by saschagrunert — COMMENTED** [review]
Honestly spoken I’m not in favor of having two dedicated packages. But I’m happy to change my opinion if this is really necessary.

**Comment by dredozubov** [comment]
I agree it may look kinda awkward, but helm 2 and 3 are far from compatible and I'd argue that a significant amount of projects have yet to switch to helm 3, so it would be helpful to have the last revision of 2.* beside the primary 3.* for a time being.

File: pkgs/applications/networking/cluster/helm2/default.nix [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  src = fetchFromGitHub {
+    owner = "helm";
+    repo = "helm";
+    rev = "v${version}";
+    sha256 = "16hbwmgq14g28r9s0ipnpiqlppyh57yrcqcspmj05vrf9jsg5vwj";
+  };
+  modSha256 = "0618zzi4x37ahsrazsr82anghhfva8yaryzb3p5d737p3ixbiyv8";
+
+  subPackages = [ "cmd/helm" ];
+  buildFlagsArray = [ "-ldflags=-w -s -X helm.sh/helm/v3/internal/version.version=v${version}" ];
```
Comment (Frostman) [inline pkgs/applications/networking/cluster/helm2/default.nix]: I believe it should be different for helm v2 - https://github.com/helm/helm/blob/release-2.16/versioning.mk#L25-L26

File: pkgs/applications/networking/cluster/helm2/default.nix [hunk 2]
```diff
@@ -0,0 +1,33 @@
+{ stdenv, buildGoModule, fetchFromGitHub, installShellFiles, Security }:
+
+buildGoModule rec {
+  pname = "helm";
+  version = "2.16.3";
```
Comment (Frostman) [inline pkgs/applications/networking/cluster/helm2/default.nix]: Could you please update it to 2.16.4 as it was released few days ago - https://github.com/helm/helm/releases/tag/v2.16.4

**Review by Frostman — CHANGES_REQUESTED** [review]
IMO it makes sense to have helm2 supported in addition to helm3 as helm2 is still looks more popular and many people using it.

File: pkgs/applications/networking/cluster/helm2/default.nix (hunk 1 above)
Comment (dredozubov) [reply pkgs/applications/networking/cluster/helm2/default.nix]: I've ported this one from the nixpkgs git history. Helm 2 was built this way before, but it probably makes sense to update it as well as the version.

File: pkgs/applications/networking/cluster/helm2/default.nix (hunk 2 above)
Comment (dredozubov) [reply pkgs/applications/networking/cluster/helm2/default.nix]: Good catch, I never expected there will be a backport to helm 2 again

Comment (Frostman) [reply pkgs/applications/networking/cluster/helm2/default.nix:5]: It's actually one more released already :) https://github.com/helm/helm/releases/tag/v2.16.5

**Review by saschagrunert — APPROVED** [review]
LGTM

File: pkgs/applications/networking/cluster/helm2/default.nix (hunk 1 above)
Comment (Frostman) [reply pkgs/applications/networking/cluster/helm2/default.nix:16]: I've tested and seems like it works correctly, so, we can keep as it was before.

**Comment by dredozubov** [comment]
I've rebased the PR against master and updated the helm 2 version to 2.16.5

**Comment by dredozubov** [comment]
Rebased, all comments addressed.

**Comment by lsowen** [comment]
@dredozubov thank you for taking this on, I for one would really like to have helm2 available, since most of our infrastructure still relies on helm2/tiller.

However, I'm not sure this change will work properly?  If you look at the old helm2 build (https://github.com/NixOS/nixpkgs/blob/a038ae1118a676f956c9cf50b78e6e0144c44266/pkgs/applications/networking/cluster/helm/default.nix), it used `buildGoPackage`, and built `tiller` and `rudder` in addition to `helm`.

**Comment by dredozubov** [comment]
@isowen You're absolutely right. I tried using `kubernetes-helm2` yesterday and it's currently broken. I haven't figured out how to properly build it yet, but I plan to work on it when I'll have the time.

**Review by saschagrunert — APPROVED** [review]
LGTM

**Comment by teto** [comment]
I am new to helm but I've already hit the helm 2/3 schism. The ecosystem warrants to keep both so +1 from me. Could you solve conflicts ?

**Comment by baracoder** [comment]
Ping
I would also be interested in having both version 2 and 3 in nixpkgs

**Comment by dredozubov** [comment]
I've checked this old branch and it's definitely broken for Helm 2. I have neither the knowledge of older golang build tools nor the time to tackle it right now. Is anyone willing to take a shot at this?

File: pkgs/applications/networking/cluster/helm2/default.nix [hunk 3]
```diff
@@ -0,0 +1,33 @@
+{ stdenv, buildGoModule, fetchFromGitHub, installShellFiles, Security }:
```
Comment (teto) [inline pkgs/applications/networking/cluster/helm2/default.nix:1]: `Security ? null` else it fails on linux

**Comment by teto** [comment]
@dredozubov  what kind of error have you encountered ? Maybe we can rebase and merge and fix that afterwards ?
What I've noticed is that the kubernetes 2 doesn't seem to have an `init` command ? (yet it returns v2.16.5). I dont know much about helm but that seemed fishy.

File: pkgs/applications/networking/cluster/helm2/default.nix [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  pname = "helm";
+  version = "2.16.5";
+
+  src = fetchFromGitHub {
+    owner = "helm";
+    repo = "helm";
+    rev = "v${version}";
+    sha256 = "16hbwmgq14g28r9s0ipnpiqlppyh57yrcqcspmj05vrf9jsg5vwj";
+  };
+  modSha256 = "0618zzi4x37ahsrazsr82anghhfva8yaryzb3p5d737p3ixbiyv8";
```
Comment (teto) [inline pkgs/applications/networking/cluster/helm2/default.nix:13]: this is the same hash as for the v3 version there is something wrong here.

File: pkgs/top-level/all-packages.nix [hunk 5]
```diff
@@ -20249,7 +20249,11 @@ in
 
   linkerd = callPackage ../applications/networking/cluster/linkerd { };
 
-  kubernetes-helm = callPackage ../applications/networking/cluster/helm { };
+  kubernetes-helm = kubernetes-helm3;
+
+  kubernetes-helm2 = callPackage ../applications/networking/cluster/helm2 { };
+
+  kubernetes-helm3 = callPackage ../applications/networking/cluster/helm3 { };
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:20256]:
```suggestion
  kubernetes-helm2 = callPackage ../applications/networking/cluster/helm2 { };
  kubernetes-helm3 = callPackage ../applications/networking/cluster/helm3 { };
```

And please move the alias to aliases.nox.

**Pull request closed without merging**